notebooks/name-search-utility-demo.ipynb

```diff
@@ -0,0 +1,331 @@
+    "    20971522: {'to': 20971524, 'Length': 16.0},\n",
+    "    ...\n",
+    "}\n",
+    "```\n",
+    "We can leave the proof for later, but this represents a directed acyclic graph representation of the stream network that we can use to figure out where the network goes. In the broarder repository, we have a whole pile of tools related to this concept and perhaps we can later come back here and make sure they are all in sync. For now, this is sufficient to mention as a teaser."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# NOTE: This is essentially the same thing provided by `get_down_connections` in networkbuilder\n",
+    "# TODO: Update networkbuilder to use a dataframe as is done here.\n",
+    "downstream_col = \"to\"\n",
+    "length_col = \"Length\"\n",
+    "\n",
+    "mask_set = set(rows.index.values)\n",
+    "rows_dl = rows[\n",
+    "    [downstream_col, length_col]\n",
+    "].copy()  # Reduce the dataframe to just the needed keys\n",
+    "connections = (rows_dl.loc[mask_set]).to_dict(\"index\")\n",
+    "if len(connections) == len(rows):\n",
+    "    print(f\"success!! -- created dictionary DAG for nwm segments.\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Now, to process\n",
+    "### 5a. Let's remind ourselves where we headed \n",
+    "_Our Goal?_\n",
+    "  * Find named segments tributary to the a particular named River\n",
+    "\n",
+    "_The general process, i.e., pseudocode?_\n",
+    "  * Make a set of named segments for the river\n",
+    "  * Make a set of connections pointing to the named segments\n",
+    "  * Drop from that set connections that are in the river itself (i.e., segments in the river that point to the next downstream segment in the river)\n",
+    "  * Optionally, drop from that set connections that don't have a name themselves\n",
+    "\n",
+    "Before we continue, \n",
+    "### 5b. We should review what we have\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(f\"Using names table: {json_names_path}\")\n",
+    "print(f\"Number of names in names table: {len(names)}\")\n",
+    "print(f\"Number of unique nwm segments with names in table: {len(ids_wnames)}\")\n",
+    "\n",
+    "print(f\"Using nwm route-link file: {full_nc_path}\")\n",
+    "print(f\"Total number of nwm segments: {len(rows)}\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 5c. Choose a River"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "name_to_find = \"Cache la Poudre River\"\n",
+    "try:\n",
+    "    named_segs = names[name_to_find]\n",
+    "except:\n",
+    "    print(f\"{name_to_find} not found in the names table\")\n",
+    "    raise ValueError\n",
```

Comment:
It seems this try/except block is unnecessary if you want to propagate the KeyError.